src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/FunctionProbes/FunctionProbesManager.cs

```diff
@@ -26,49 +29,197 @@ internal sealed class FunctionProbesManager : IFunctionProbesManager
+                // actively in one of the instrumented methods and it happens to fault.
+                //
+                return;
+            }
 
-                FunctionProbesStub.InstrumentedMethodCache = null;
-                RequestFunctionProbeUninstallation();
+            OnProbeFault?.Invoke(this, instrumentedMethod);
+        }
+        
+        private void TransitionStateFromHr(TaskCompletionSource? taskCompletionSource, int hresult, long expectedState, long succeededState, long failedState)
+        {
+            Exception? ex = Marshal.GetExceptionForHR(hresult);
+            long newState = (ex == null) ? succeededState : failedState;
 
-                _isCapturing = false;
+            if (expectedState != Interlocked.CompareExchange(ref _probeState, newState, expectedState))
+            {
+                // Unexpected, the profiler is in a different state than us.
+                StateMismatch(expectedState);
+                return;
+            }
+
+            if (ex == null)
+            {
+                _ = taskCompletionSource?.TrySetResult();
             }
+            else
+            {
+                _ = taskCompletionSource?.TrySetException(ex);
+            }
+        }
+
+        private void StateMismatch(long expected)
+        {
+            InvalidOperationException ex = new(string.Format(CultureInfo.InvariantCulture, ParameterCapturingStrings.ErrorMessage_ProbeStateMismatchFormatString, expected, _probeState));
+
+            _probeState = ProbeStateUnrecoverable;
+            _ = _probeRegistrationTaskSource.TrySetException(ex);
+            _ = _installationTaskSource?.TrySetException(ex);
+            _ = _uninstallationTaskSource?.TrySetException(ex);
         }
 
-        public void StartCapturing(IList<MethodInfo> methods)
+        public async Task StopCapturingAsync(CancellationToken token)
         {
+            DisposableHelper.ThrowIfDisposed<FunctionProbesManager>(ref _disposedState);
+
+            if (ProbeStateInstalled != Interlocked.CompareExchange(ref _probeState, ProbeStateUninstalling, ProbeStateInstalled))
+            {
+                throw new InvalidOperationException(string.Format(CultureInfo.InvariantCulture, ParameterCapturingStrings.ErrorMessage_ProbeStateMismatchFormatString, ProbeStateInstalled, _probeState));
+            }
+
+            _uninstallationTaskSource = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
+            try
+            {
+                StopCapturingCore();
+            }
+            catch
+            {
+                _uninstallationTaskSource = null;
+                _probeState = ProbeStateInstalled;
+                throw;
+            }
+
+            using CancellationTokenSource cts = CancellationTokenSource.CreateLinkedTokenSource(_disposalTokenSource.Token, token);
+            using IDisposable _ = cts.Token.Register(() =>
```

Comment:
What's the behavior here if we are already disposed? Will the token registration action run immediately?